File: ios/Runner/MovieInfoProvider.swift
import Foundation

protocol MovieInfoProvider {
     func provideInfo(outputDir: String) -> MovieInfo?
}

class MovieInfoProviderImpl : MovieInfoProvider {
    func provideInfo(outputDir: String) -> MovieInfo? {
        var uuid = self.generateUUID()
        let files: [String]
        
        do {
            files = try FileManager.default.contentsOfDirectory(atPath: outputDir)
        } catch {
            return nil
        }
        
        while files.any(predicate: { (file:String) -> Bool in
            file.contains(uuid)
        }) {
             uuid = self.generateUUID()
        }
        
        let path = "\(outputDir)/\(uuid)"
        return MovieInfo(thumbPath: "\(path).jpg", moviePath: "\(path).mp4")
    }
    
    private func generateUUID() -> String {
        return UUID.init().uuidString
    }
}

struct MovieInfo {
    let thumbPath: String
    let moviePath: String
}

File: android/app/src/main/kotlin/com/vitoksmile/cpmoviemaker/MovieInfoProvider.kt
package com.vitoksmile.cpmoviemaker

import java.io.File
import java.util.*

interface MovieInfoProvider {
    fun provideInfo(outputDir: String): Info

    data class Info(
        val thumbPath: String,
        val moviePath: String
    )
}

class MovieInfoProviderImpl : MovieInfoProvider {
    override fun provideInfo(outputDir: String): MovieInfoProvider.Info {
        var uuid = generateUUID()
        val files = File(outputDir).listFiles().map { it.name }
        while (files.any { it.contains(uuid) }) {
            uuid = generateUUID()
        }

        return MovieInfoProvider.Info(
            thumbPath = createPath(outputDir, uuid, "jpg"),
            moviePath = createPath(outputDir, uuid, "mp4")
        )
    }

    private fun generateUUID() = UUID.randomUUID().toString()

    private fun createPath(outputDir: String, uuid: String, format: String) = buildString {
        append(outputDir)
        append(File.separator)
        append(uuid)
        append(".")
        append(format)
    }
}
